private_sharing/api_views.py

```diff
@@ -226,7 +233,50 @@ def post(self, request, format=None):
         return Response("success")
 
 
-class ProjectFileDirectUploadView(ProjectFormBaseView):
+class SaveDataTypesMixin(object):
+    """
+    Mixin to do some final verification of datatypes to be associated with the datafile
+    and to do the actual association.
+    """
+
+    @property
+    def good_datatypes(self):
+        """
+        Verifies that the supplied datatypes are valid.  Returns boolean.
+        """
+        # running this test here as we have the object available in the view, but
+        # not in the form.
```

Comment:
It _is_ possible to pass something to a form when instantiated, for example `BaseProjectMembersForm`'s `__init__` method setting a `self.project`: https://github.com/OpenHumans/open-humans/blob/master/private_sharing/forms.py#L158

The generic FormView makes this customizable with a `get_form_kwargs` method. It looks like `ProjectFormBaseView` doesn't provide something like this, so... not worth worrying about.